File: src/main/java/com/dude/dms/ui/components/tags/AttributeValueField.java
package com.dude.dms.ui.components.tags;

import com.dude.dms.backend.brain.DmsLogger;
import com.dude.dms.backend.data.docs.AttributeValue;
import com.dude.dms.backend.service.AttributeValueService;
import com.vaadin.flow.component.datepicker.DatePicker;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.textfield.NumberField;
import com.vaadin.flow.component.textfield.TextField;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.util.function.Function;

public class AttributeValueField extends HorizontalLayout {

    private static final DmsLogger LOGGER = DmsLogger.getLogger(AttributeValueField.class);

    private final Function<Void, Boolean> validation;

    private final String label;

    private final boolean required;

    public AttributeValueField(AttributeValue attributeValue, AttributeValueService attributeValueService) {
        label = attributeValue.getAttribute().getName();
        required = attributeValue.getAttribute().isRequired();
        switch (attributeValue.getAttribute().getType()) {
            case STRING:
                TextField textField = new TextField(label);
                if (attributeValue.getStringValue() != null) {
                    textField.setValue(attributeValue.getStringValue());
                }
                textField.setWidthFull();
                textField.addValueChangeListener(event -> {
                    if (event.getHasValue().isEmpty() && required) {
                        LOGGER.showError("Attribute '" + label + "' can not be empty.");
                    } else {
                        attributeValue.setStringValue(event.getValue());
                        attributeValueService.save(attributeValue);
                        LOGGER.showInfo("Saved '" + label + "'!");
                    }
                });
                add(textField);
                validation = v -> (!required || !textField.isEmpty());
                break;
            case INT:
                NumberField intField = new NumberField(label);
                intField.setStep(1);
                if (attributeValue.getIntValue() != null) {
                    intField.setValue(Double.valueOf(attributeValue.getIntValue()));
                }
                intField.setWidthFull();
                intField.addValueChangeListener(event -> {
                    if (event.getHasValue().isEmpty() && required) {
                        LOGGER.showError("Attribute '" + label + "' can not be empty.");
                    } else {
                        attributeValue.setIntValue(event.getValue().intValue());
                        attributeValueService.save(attributeValue);
                        LOGGER.showInfo("Saved '" + label + "'!");
                    }
                });
                add(intField);
                validation = v -> (!required || !intField.isEmpty());
                break;
            case FLOAT:
                NumberField floatField = new NumberField(label);
                floatField.setStep(0.01);
                if (attributeValue.getFloatValue() != null) {
                    floatField.setValue(Double.valueOf(attributeValue.getFloatValue()));
                }
                floatField.setWidthFull();
                floatField.addValueChangeListener(event -> {
                    if (event.getHasValue().isEmpty() && required) {
                        LOGGER.showError("Attribute '" + label + "' can not be empty.");
                    } else {
                        attributeValue.setFloatValue(event.getValue().floatValue());
                        attributeValueService.save(attributeValue);
                        LOGGER.showInfo("Saved '" + label + "'!");
                    }
                });
                add(floatField);
                validation = v -> (!required || !floatField.isEmpty());
                break;
            case DATE:
                DatePicker datePicker = new DatePicker(label);
                if (attributeValue.getDateValue() != null) {
                    datePicker.setValue(attributeValue.getDateValue());
                }
                datePicker.setWidthFull();
                datePicker.addValueChangeListener(event -> {
                    if (event.getHasValue().isEmpty() && required) {
                        LOGGER.showError("Attribute '" + label + "' can not be empty.");
                    } else {
                        attributeValue.setDateValue(event.getValue());
                        attributeValueService.save(attributeValue);
                        LOGGER.showInfo("Saved '" + label + "'!");
                    }
                });
                add(datePicker);
                validation = v -> (!required || !datePicker.isEmpty());
                break;
            default:
                throw new NotImplementedException();
        }
    }

    public boolean validate() {
        return validation.apply(null);
    }

    public String getLabel() {
        return label;
    }

    public boolean isRequired() {
        return required;
    }
}

File: src/main/java/com/dude/dms/ui/components/tags/AttributeValueField.kt
package com.dude.dms.ui.components.tags

import com.dude.dms.backend.data.docs.Attribute
import com.dude.dms.backend.data.docs.AttributeValue
import com.dude.dms.extensions.attributeValueService
import com.dude.dms.extensions.findDate
import com.dude.dms.extensions.findDecimal
import com.dude.dms.extensions.round
import com.dude.dms.ui.components.misc.DocImageEditor
import com.vaadin.flow.component.button.Button
import com.vaadin.flow.component.button.ButtonVariant
import com.vaadin.flow.component.combobox.ComboBox
import com.vaadin.flow.component.datepicker.DatePicker
import com.vaadin.flow.component.icon.VaadinIcon
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.component.textfield.NumberField

class AttributeValueField(val attributeValue: AttributeValue, imageEditor: DocImageEditor? = null) : HorizontalLayout() {

    private val validation: () -> Boolean

    private val isRequired = attributeValue.attribute.isRequired

    val label = attributeValue.attribute.name + if (isRequired) " *" else ""

    private val textField
        get() = ComboBox<String>(label).apply {
            isAllowCustomValue = true
            val items = attributeValueService.findAutocomplete(attributeValue.attribute)
            if (items.isNotEmpty()) {
                setItems()
            }
            value = attributeValue.stringValue ?: ""
            addValueChangeListener { event ->
                if (!event.hasValue.isEmpty || !isRequired) {
                    attributeValue.stringValue = event.value
                    if (attributeValue.doc != null) attributeValueService.save(attributeValue)
                }
            }
            addCustomValueSetListener { value = it.detail }
            setWidthFull()
        }

    private val intField
        get() = NumberField(label).apply {
            step = 1.0
            value = attributeValue.intValue?.toDouble()
            addValueChangeListener { event ->
                if (!event.hasValue.isEmpty || !isRequired) {
                    attributeValue.intValue = event.value.toInt()
                    if (attributeValue.doc != null) attributeValueService.save(attributeValue)
                }
            }
            setWidthFull()
        }

    private val floatField
        get() = NumberField(label).apply {
            step = 0.01
            value = attributeValue.floatValue?.toDouble()?.round(2)
            addValueChangeListener { event ->
                if (!event.hasValue.isEmpty || !isRequired) {
                    attributeValue.floatValue = event.value.toFloat()
                    if (attributeValue.doc != null) attributeValueService.save(attributeValue)
                }
            }
            setWidthFull()
        }

    private val datePicker
        get() = DatePicker(label).apply {
            attributeValue.dateValue?.let { value = it }
            addValueChangeListener { event ->
                if (!event.hasValue.isEmpty || !isRequired) {
                    attributeValue.dateValue = event.value
                    if (attributeValue.doc != null) attributeValueService.save(attributeValue)
                }
            }
            setWidthFull()
        }

    init {
        isSpacing = false
        when (attributeValue.attribute.type) {
            Attribute.Type.STRING -> {
                val field = textField
                add(field)
                validation = { !isRequired || (!field.isEmpty && !field.value.isNullOrBlank()) }
                if (imageEditor != null) {
                    val pick = Button(VaadinIcon.CROSSHAIRS.create()) { event ->
                        imageEditor.pickEvent = {
                            field.value = it?.word?.text ?: field.value
                            event.source.removeThemeVariants(ButtonVariant.LUMO_SUCCESS)
                        }
                        event.source.addThemeVariants(ButtonVariant.LUMO_SUCCESS)
                    }.apply { style["margin"] = "auto auto 5px auto" }
                    add(pick)
                }
            }
            Attribute.Type.INT -> {
                val field = intField
                add(field)
                validation = { !isRequired || !field.isEmpty }
                if (imageEditor != null) {
                    val pick = Button(VaadinIcon.CROSSHAIRS.create()) { event ->
                        imageEditor.pickEvent = {
                            field.value = it?.word?.text?.toDoubleOrNull() ?: field.value
                            event.source.removeThemeVariants(ButtonVariant.LUMO_SUCCESS)
                        }
                        event.source.addThemeVariants(ButtonVariant.LUMO_SUCCESS)
                    }.apply { style["margin"] = "auto auto 5px auto" }
                    add(pick)
                }
            }
            Attribute.Type.FLOAT -> {
                val field = floatField
                add(field)
                validation = { !isRequired || !field.isEmpty }
                if (imageEditor != null) {
                    val pick = Button(VaadinIcon.CROSSHAIRS.create()) { event ->
                        imageEditor.pickEvent = {
                            field.value = it?.word?.text?.findDecimal() ?: field.value
                            event.source.removeThemeVariants(ButtonVariant.LUMO_SUCCESS)
                        }
                        event.source.addThemeVariants(ButtonVariant.LUMO_SUCCESS)
                    }.apply { style["margin"] = "auto auto 5px auto" }
                    add(pick)
                }
            }
            Attribute.Type.DATE -> {
                val field = datePicker
                add(field)
                validation = { !isRequired || !field.isEmpty }
                if (imageEditor != null) {
                    val pick = Button(VaadinIcon.CROSSHAIRS.create()) { event ->
                        imageEditor.pickEvent = {
                            field.value = it?.word?.text?.findDate() ?: field.value
                            event.source.removeThemeVariants(ButtonVariant.LUMO_SUCCESS)
                        }
                        event.source.addThemeVariants(ButtonVariant.LUMO_SUCCESS)
                    }.apply { style["margin"] = "auto auto 5px auto" }
                    add(pick)
                }
            }
        }
    }

    fun validate() = validation.invoke()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as AttributeValueField
        if (attributeValue != other.attributeValue) return false
        return true
    }

    override fun hashCode() = attributeValue.hashCode()
}